src/realm/index_string.cpp

```diff
@@ -197,38 +197,39 @@ size_t IndexArray::from_list<index_FindAll_nocopy>(StringData value, IntegerColu
     return size_t(FindRes_column);
 }
 
-template <IndexMethod method, class T>
+template <IndexMethod method>
 size_t IndexArray::index_string(StringData value, IntegerColumn& result, InternalFindResult& result_ref,
                                 ColumnBase* column) const
 {
     // Return`realm::not_found`, or an index to the (any) match
-    bool first(method == index_FindFirst);
+    constexpr bool first(method == index_FindFirst);
     // Return 0, or the number of items that match the specified `value`
-    bool get_count(method == index_Count);
+    constexpr bool get_count(method == index_Count);
     // Place all row indexes containing `value` into `result`
     // Returns one of FindRes_not_found[==0] if no matches found
     // Returns FindRes_single, if one match found: the result row literal is
     // both placed in `result_ref.payload` and added to `column`
     // Returns FindRes_column, if more than one match found: the matching row
     // literals are copied into `column`
-    bool all(method == index_FindAll);
+    constexpr bool all(method == index_FindAll);
     // Same as `index_FindAll` but does not copy matching rows into `column`
     // returns FindRes_not_found if there are no matches
     // returns FindRes_single and the row index (literal) in result_ref.payload
     // or returns FindRes_column and the reference to a column of duplicates in
     // result_ref.result with the results in the bounds start_ndx, and end_ndx
-    bool allnocopy(method == index_FindAll_nocopy);
+    constexpr bool allnocopy(method == index_FindAll_nocopy);
+
+    constexpr size_t local_not_found = (allnocopy || all) ? size_t(FindRes_not_found) : first ? not_found : 0;
```

Comment:
Ok I understand now. Thanks for explaining!
To convince myself, I also checked the generated assembly. There are indeed no compares in sight, just moving constants into registers 👍 
```
 24906         movb    $1, -213(%rbp)  #, first
 24907         .loc 31 207 0
 24908         movb    $0, -212(%rbp)  #, get_count
 24909         .loc 31 214 0
 24910         movb    $0, -211(%rbp)  #, all
 24911         .loc 31 220 0
 24912         movb    $0, -210(%rbp)  #, allnocopy
 24913         .loc 31 222 0
 24914         movq    $-1, -184(%rbp) #, local_not_found
```